### DISCUSSION TITLE
Matlab: DIPimage shadows Matlab toolbox functions

### DISCUSSION THREAD

**Discussion by AndreZeug** [DIPlib/diplib#174] [General] [answered]
I am using DIPimage 3.x within Matlab R2024a and want to use the Matlab Toolbox function "cluster" to do hierarchical clustering of multidimensional data. Unfortunately the toolbox function is shadowed by the DIPimage toolbox.
```
>> which -all cluster
Z:\GIT\DIPlib\share\DIPimage\cluster.m
C:\Program Files\MATLAB\R2024a\toolbox\stats\stats\cluster.m                  % Shadowed
C:\Program Files\MATLAB\R2024a\toolbox\bioinfo\bioinfo\@phytree\cluster.m     % phytree method
C:\Program Files\MATLAB\R2024a\toolbox\stats\stats\@gmdistribution\cluster.m  % gmdistribution method
```
Is there an easy way to call such shadowed function?

**Comment by crisluengo** [comment, marked as answer]
Take a look at the answer at the bottom here: https://diplib.org/diplib-docs/sec_dum_faq.html#sec_dum_faq_usage_select

Name clashes are inevitable in a language without namespaces, which MATLAB was when we created DIPimage. We avoided using names of existing functions in the toolboxes most relevant to image processing, but that's a moving target, they keep adding new functions and new toolboxes. MATLAB does have namespaces nowadays, but adding a namespace to the DIPimage toolbox is akin to changing all function names, which we don't want to do for backwards-compatibility.

So the only thing we can do is show how to call a shadowed function: change directories first:
```matlab
old_dir = cd('C:\Program Files\MATLAB\R2024a\toolbox\stats\stats')
cluster(...)
cd(old_dir)
```

For generic code you need to find that path name first, which involves calling `which` as a function, getting its output cell array, then iterating over the results and see which one has "toolbox\stats\stats" in the path (and note that on Linux or macOS the path will have forward slashes, so use `fullfile('toolbox','stats','stats')` in your search).

Sorry, this is not the most beautiful solution. :(

**Comment by AndreZeug** [comment]
Dear Cris, thank you so much for your prompt reply. 
Even though this is not the most elegant way, it is straightforward and solves my problem.
Just great!
Andre
